pkg/sync/sync.go

```diff
@@ -309,74 +366,99 @@ func copyInParallel(src, dst object.ObjectStorage, obj object.Object) error {
-	deleteObj := func(storage object.ObjectStorage, key string, start time.Time, dry bool) {
-		if dry {
-			logger.Infof("Will delete %s from %s", key, storage)
-			return
-		}
-		if err := try(3, func() error { return storage.Delete(key) }); err == nil {
-			deleted.Increment()
-			logger.Debugf("Deleted %s from %s in %s", key, storage, time.Since(start))
-		} else {
-			failed.Increment()
-			logger.Errorf("Failed to delete %s from %s: %s", key, storage, err)
+func copyData(src, dst object.ObjectStorage, key string, size int64) error {
+	start := time.Now()
+	var multiple bool
+	var err error
+	if size < maxBlock {
+		err = try(3, func() error { return doCopySingle(src, dst, key, size) })
+	} else {
+		var upload *object.MultipartUpload
+		if upload, err = dst.CreateMultipartUpload(key); err == nil {
+			multiple = true
+			err = doCopyMultiple(src, dst, key, size, upload)
+		} else { // fallback
+			err = try(3, func() error { return doCopySingle(src, dst, key, size) })
 		}
 	}
-	copyPerms := func(fi object.File, start time.Time, log bool) {
-		key := fi.Key()
-		if err := dst.(object.FileSystem).Chmod(key, fi.Mode()); err != nil {
-			logger.Warnf("Chmod %s to %d: %s", key, fi.Mode(), err)
-		}
-		if err := dst.(object.FileSystem).Chown(key, fi.Owner(), fi.Group()); err != nil {
-			logger.Warnf("Chown %s to (%s,%s): %s", key, fi.Owner(), fi.Group(), err)
-		}
-		if log {
-			logger.Debugf("Copied permissions (%s:%s:%s) for %s in %s", fi.Owner(), fi.Group(), fi.Mode(), key, time.Since(start))
+	if err == nil {
+		if !multiple {
+			copiedBytes.IncrInt64(size)
 		}
+		logger.Debugf("Copied data of %s (%d bytes) in %s", key, size, time.Since(start))
+	} else {
+		logger.Errorf("Failed to copy data of %s in %s: %s", key, time.Since(start), err)
 	}
+	return err
+}
 
+func worker(tasks <-chan object.Object, src, dst object.ObjectStorage, config *Config) {
 	for obj := range tasks {
-		start := time.Now()
+		key := obj.Key()
 		switch obj.Size() {
 		case markDeleteSrc:
-			deleteObj(src, obj.Key(), start, config.Dry)
+			deleteObj(src, key, config.Dry)
 		case markDeleteDst:
-			deleteObj(dst, obj.Key(), start, config.Dry)
+			deleteObj(dst, key, config.Dry)
 		case markCopyPerms:
 			if config.Dry {
-				logger.Infof("Will copy permissions for %s", obj.Key())
+				logger.Infof("Will copy permissions for %s", key)
 				break
 			}
-			copyPerms(obj.(object.File), start, true)
+			copyPerms(dst, obj)
 			copied.Increment()
+		case markChecksum:
+			if config.Dry {
+				logger.Infof("Will compare checksum for %s", key)
+				break
+			}
+			obj = obj.(*withSize).Object
+			if equal, err := checkSum(src, dst, key, obj.Size()); err == nil && equal {
+				if config.DeleteSrc {
+					deleteObj(src, key, false)
+				} else if config.Perms { // FIXME: && needCopyPerms(obj, dstobj)
+					copyPerms(dst, obj)
+					copied.Increment()
+				} else {
+					skipped.Increment()
+				}
+				break
+			}
+			// checkSum failed or not equal, copy the object
```

Comment:
For failed check, should we mark it as failed?

Reply:
Copying the object might fix the problem if error occurs in the destination side.

Reply:
Maybe it is safer to just mark it failed and let users to double check it.